Sources/PackageModel/PackageIdentity.swift

```diff
@@ -0,0 +1,230 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Foundation
+
+import TSCBasic
+import TSCUtility
+
+public struct PackageIdentity: LosslessStringConvertible {
+    public let description: String
+    public let computedName: String
+
+    public init(_ string: String) {
```

Comment:
The path part of the URL is often different between HTTP and SSH URLs for the same repository.  Where does that normalization take place?

Reply:
How do you mean? Do you have any specific examples in mind?

(Also, I pushed some changes since you started reviewing; there's a lot more explanation about what's going on as of 844286bc09ca14c7b4f7a0318bcb0b4328c6a576)

Reply:
I'm not sure if there's a public example to point to, but one instance of BitBucket Server that I know of has something that follows this pattern:
```
ssh://git@server.company.com/teamname/projectname.git
https://server.company.com/scm/teamname/projectname.git
```
I'm not sure if the extra "scm" part in the HTTP form is something that's configured on the server.  My impression is that there isn't generally a guarantee that the path part of the URLs are the same between SSH and HTTP.

Reply:
@abertelrud do you think there is a way to canonicalize the BitBucket example above?

Reply:
I'm not sure.  @neonichu do you know what the deal is with this server, is this path component something that a repository server can just randomly configure to be different for SSH vs HTTPS URLs, or is there some kind of pattern?  I looked around a bit for configuration information about that server, but haven't found anything yet.

Reply:
I have also pinged somebody here who I think would know and/or could point in the right direction.  In any case it doesn't seem as if this would block this PR, since this isn't yet actually changing the identity.

Reply:
@abertelrud Thanks for sharing that example. 

The normalization proposed here attempts to mitigate what are, effectively, cosmetic differences in URLs. There are going to be limitations to this approach, like the one you describe, where a host uses different conventions for HTTPS and SSH URLs (my understanding is that most hosts, including GitHub and GitLab, do a direct translation). We could tailor our algorithm to accommodate this case specifically (e.g. by only looking at authority + last 2 path components), but I'm hesitant to take such an approach for loss of generality.

If you wanted to reconcile this on the client-side, the most direct way would be to set a dependency mirror URL. Another approach is something @tomerd had mentioned before, where SPM could use the first commit of each repository as a "fingerprint" to identify the same repositories at different URLs (e.g. `/MorningStar.git` and `/EveningStar.git`)

Reply:
I agree that we shouldn't add special cases here, and I've confirmed that this is configurable on the server, so there isn't really anything that could be added that would cover all cases.  So I guess my main concern is that we will have cases that stop working because of this, and that it might not be apparent how to address it.

If we can fetch both repositories and compare the first-commit hashes then that might mitigate this, although it would still mean needing to have credentials for both repositories so they can be fetched (whereas today only the overriding repository needs credentials).

Reply:
This also won't deal with any Git remappings that might have been set up in `git config`, though I don't know how common that is in practice.

Reply:
IIRC this was discussed quite a bit a few years ago with the conclusion that it would break too many cases.  But I don't have the context to know whether that was the right conclusion.

Reply:
@abertelrud Yeah, too many unknowns to know for sure what the actual impact will be. We can / should gate the new semantics behind a feature flag / command line option. This will help us identify edge cases without breaking anything in the short term.